Simplify setup instructions and guide (#2970)
rubyforgood/human-essentials: README.md

#### 💬 Join us on slack
Please feel free to join us on Slack!
You can sign up at https://rubyforgood.herokuapp.com and find us in #human-essentials.

There are numerous other folks that can chime in and answer questions -- please ask and someone will probably be there to help!

## 🛠️ Getting Started

### Install Ruby
You will need to first install the required ruby version specified in the `.ruby-version` file. GoRails has a very detail oriented guide for installing ruby on Ubuntu, Windows, and macOSX. You can check out that guide [here](https://gorails.com/setup/osx/12-monterey). **Follow only the Installing Ruby step, as our project setup differs**

It is highly recommended you use a ruby version manager such as:
- [rbenv](https://github.com/rbenv/rbenv) 
- [asdf](https://asdf-vm.com/)
- [rvm](https://rvm.io/)

You can verify that your ruby installation was successful and matches the version in `.ruby-version`  in the project directory:
```
ruby -v
```

Once you've successfully installed ruby. You may proceed to the next section!

#### Troubleshooting

> The rbenv repository provides a [rbenv-doctor script](https://github.com/rbenv/rbenv-installer#rbenv-doctor) to verify the state of the rbenv installation and if a ruby version is installed

Author: ```suggestion
**My RBENV installation did not work**
The rbenv repository provides a [rbenv-doctor script](https://github.com/rbenv/rbenv-installer#rbenv-doctor) to verify the state of the rbenv installation and if a ruby version is installed
```

 - Check the [`CONTRIBUTING.md`](CONTRIBUTING.md) file for a guide on how to get started
 - This is a 100% volunteer-supported project, please be patient with your correspondence. We do handle issues and PRs with more fervor during Hacktoberfest & Conferences, but most (all?) of us have day jobs and so responses to questions / pending PRs may not be immediate. Please be patient, we'll get to you! :)

#### 💬 Join us on slack
Please feel free to join us on Slack!
You can sign up at https://rubyforgood.herokuapp.com and find us in #human-essentials.

There are numerous other folks that can chime in and answer questions -- please ask and someone will probably be there to help!

## 🛠️ Getting Started

### Install Ruby
You will need to first install the required ruby version specified in the [`.ruby-version`](.ruby-version) file. GoRails has a very detail oriented guide for installing ruby on Ubuntu, Windows, and macOSX. You can check out that guide [here](https://gorails.com/setup/osx/12-monterey). **Follow only the Installing Ruby step, as our project setup differs**

> It is highly recommended you use a ruby version manager such as:

Reviewer: I wonder if this section isn't a bit confusing to newer folks. The gorails guide above has you install rbenv above. So if you're new and following the guide above, you're fine and this irrelevant information. If you're experienced you understand the importance of a version manager and already have one that you like and so pointing out options probably isn't necessary. 

Additionally, if I'm new you've given me three options, but not told me how to make a choice. At a minimum, it would be good to say that rbenv is the preferred option or the one that should be picked unless you have a reason to pick the other ones. 

Removing this section completely makes the guide slightly shorter, and I don't think it makes it any less informative.

```
node -v
```

*NOTE: It's possible that Node version 12 may cause you problems, see issue #751. Node 10 or 11 seem to be fine.*

### Install Yarn
You can install yarn by following the instructions [here](https://classic.yarnpkg.com/lang/en/docs/install/#mac-stable)

You can verify that yarn was installed correctly by running this and seeing a version get returned in the project directory:
```
yarn -v
```

### Install Postgres
You must install postgres and run the database locally. Instructions differ depending on the operating system.

Follow one of these guides to install postgres:
This app uses PostgreSQL for all environments. You'll also need to create the `dev` and `test` databases, the app is expecting them to be named `diaper_dev` and `diaper_test` respectively. This should all be handled with `rails db:setup`.
Create a `database.yml` file on `config/` directory with your database configurations. You can also copy the existing file called [`database.yml.example`](config/database.yml.example) as an example and just change the credentials.

> >>>>>>> d97e5595a6390ad27c164c1b85ccd6444272287c

Author: ```suggestion
```

<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
[![All Contributors](https://img.shields.io/badge/all_contributors-111-orange.svg?style=flat-square)](#contributors-)
<!-- ALL-CONTRIBUTORS-BADGE:END -->

[![View performance data on Skylight](https://badges.skylight.io/status/LrXHcxDK7Be9.svg)](https://oss.skylight.io/app/applications/LrXHcxDK7Be9)
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Frubyforgood%2Fdiaper.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2Frubyforgood%2Fdiaper?ref=badge_shield)
[![Knapsack Pro Parallel CI builds for RSpec tests](https://img.shields.io/badge/Knapsack%20Pro-Parallel%20%2F%20Rspec%20tests-%230074ff)](https://knapsackpro.com/dashboard/organizations/1858/projects/1295/test_suites/1835/builds?utm_campaign=organization-id-1858&amp;utm_content=test-suite-id-1835&amp;utm_medium=readme&amp;utm_source=knapsack-pro-badge&amp;utm_term=project-id-1295)

## 💖 Mission

> Human Essentials is an inventory management system that is built to address the needs of [Diaper Banks](https://nationaldiaperbanknetwork.org/diaper-need/) as directly and explicitly as possible, and later adapted to meet the need of other Essentials Banks. Essentials Banks maintain inventory, receive donations and other means of intaking human essentials supplies (e.g. diapers, period supplies), and issue distributions to community partner organizations. Like any non-profit, they also need to perform reports on this data, and have day-to-day operational information they need as well. This application aims to serve all those needs, as well as facilitate, wherever possible the general operations of the Diaper Bank themselves (eg. through using barcode readers, scale weighing, inventory audits).

Reviewer 2: is an inventory management system that *was* built

> Human Essentials is an inventory management system that is built to address the needs of [Diaper Banks](https://nationaldiaperbanknetwork.org/diaper-need/) as directly and explicitly as possible, and later adapted to meet the need of other Essentials Banks. Essentials Banks maintain inventory, receive donations and other means of intaking human essentials supplies (e.g. diapers, period supplies), and issue distributions to community partner organizations. Like any non-profit, they also need to perform reports on this data, and have day-to-day operational information they need as well. This application aims to serve all those needs, as well as facilitate, wherever possible the general operations of the Diaper Bank themselves (eg. through using barcode readers, scale weighing, inventory audits).

Author: ```suggestion
Human Essentials is an inventory management system that was built to address the needs of [Diaper Banks](https://nationaldiaperbanknetwork.org/diaper-need/) as directly and explicitly as possible, and later adapted to meet the need of other Essentials Banks. Essentials Banks maintain inventory, receive donations and other means of intaking human essentials supplies (e.g. diapers, period supplies), and issue distributions to community partner organizations. Like any non-profit, they also need to perform reports on this data, and have day-to-day operational information they need as well. This application aims to serve all those needs, as well as facilitate, wherever possible the general operations of the Diaper Bank themselves (eg. through using barcode readers, scale weighing, inventory audits).
```

## 🌟 Impact

> Human Essentials serves over X organizations across the United States at **no cost to them**. It has managed Y number of distributions to Z number of families. Our team is in partnership with the [National Diaper Bank Network (NDBN)](https://nationaldiaperbanknetwork.org/) and can be found in their annual conference that brings T number of non-profit organizations that distribute essential products to people.

Author: ```suggestion
Human Essentials has over 200 registered banks across the United States at **no cost to them**. It is currently helping over **3 million children** receive diapers and over **400k period supply recipients** receive period supplies. Our team is in partnership with the [National Diaper Bank Network (NDBN)](https://nationaldiaperbanknetwork.org/) and can be found in their annual conference that brings T number of non-profit organizations that distribute essential products to people.
```

> Human Essentials serves over X organizations across the United States at **no cost to them**. It has managed Y number of distributions to Z number of families. Our team is in partnership with the [National Diaper Bank Network (NDBN)](https://nationaldiaperbanknetwork.org/) and can be found in their annual conference that brings T number of non-profit organizations that distribute essential products to people.

Author: @seanmarcia could you replace the spots with metrics?

Reviewer 2: Human essentials has over 200 registered banks across...

It has managed Y number of distributions to Z number of families. -> It is currently helping over 3 million children receive diapers and over 400k period supply recipients receive period supplies.

 - Check the `CONTRIBUTING.md` file for a guide on how to get started
 - This is a 100% volunteer-supported project, please be patient with your correspondence. We do handle issues and PRs with more fervor during Hacktoberfest & Conferences, but most (all?) of us have day jobs and so responses to questions / pending PRs may not be immediate. Please be patient, we'll get to you! :)

#### 💬 Join us on slack
Please feel free to join us on Slack!
You can sign up at https://rubyforgood.herokuapp.com and find us in #human-essentials.

There are numerous other folks that can chime in and answer questions -- please ask and someone will probably be there to help!

## 🛠️ Getting Started

### Install Ruby

> You will need to first install the required ruby version specified in the `.ruby-version` file.  It is highly recommended you use a ruby version manager like [rbenv](https://github.com/rbenv/rbenv) or [asdf](https://asdf-vm.com/)

Reviewer 3: Did a clean install of MacOS 12.4 with an Intel x86_64 MacBook. 
I had a problem after installing `rbenv` via `homebrew` where the shims folder was not setup. [This solved it](https://stackoverflow.com/questions/72140251/how-to-set-path-correctly-from-rbenv-shims-on-mac-m1-with-zsh)

Author: Did the section on installing ruby in GoRails below solve the issue?

Reviewer 3: I didn't check the GoRails guide 🙈 
I'll take a look and share my findings

GoRails has a very detail oriented guide for installing ruby on Ubuntu, Windows, and macOSX. You can check out that guide [here](https://gorails.com/setup/osx/12-monterey). **Follow only the Installing Ruby step, as our project setup differs**

Once you've succesfully installed ruby. You may proceed to the next section!

### Install Node

> You must install the version of node specified in `.nvmrc`. It is highly recommended to use [nvm](https://github.com/nvm-sh/nvm).

Reviewer 3: I'd try to be more specific (Add the step to run `nvm install` ?)

Author: I do like the idea of being more specific but the challenge is keeping it up to date. I found linking to other sources that are up to date is a little more future-proof.

Reviewer 3: My train of thought / main concern here is trying to provide context to a person who has never seen a `.nvmrc` file before. 

I'd add a link to the file in the repo so they can see is just a plain text file specifying the version. The part I'm struggling with is how to explain that you can run `nvm install` in the project folder and it should fetch the specified version or you can manually `run nvm install version`

Reviewer 4: @heyapricot I think you did very well yourself. 'You can run nvm install in the project folder and it should fetch the specified version or you can manually run nvm install version'. That was very clear to me.

I don't know how often the version of '.nvmrc' changes, but I think it might be useful to just write it here. Something like: 
Version 12.14.1 (Last review 2022.6.15)

Author: > @heyapricot I think you did very well yourself. 'You can run nvm install in the project folder and it should fetch the specified version or you can manually run nvm install version'. That was very clear to me.

Will be adding! Thanks!

> I don't know how often the version of '.nvmrc' changes, but I think it might be useful to just write it here. Something like:

You are right node hardly changes and it is a relatively quick fix if it goes out of date. I can add that in.

You can follow the guide [here](https://heynode.com/tutorial/install-nodejs-locally-nvm/) for setting up node.

*NOTE: It's possible that Node version 12 may cause you problems, see issue #751. Node 10 or 11 seem to be fine.*

### Install Postgres
You must install postgres and run the database locally. Instructions differ depending on the operating system.

Follow one of these guides to install postgres:

###### MacOSX

> https://www.digitalocean.com/community/tutorials/how-to-use-postgresql-with-your-ruby-on-rails-application-on-macos

Reviewer 3: I ran the `bin/setup` script just after installing postgresql via homebrew and got the following error:

```ruby
ActiveRecord::DatabaseConnectionError: There is an issue connecting to your database with your username/password, username: postgres.

Caused by:
PG::ConnectionBad: connection to server at "::1", port 5432 failed: FATAL:  role "postgres" does not exist
```

Homebrew runs initdb after the postgresql installation. I had a role created for my user but no database with the name of my username so I couldn't use psql.
I had to create the `postgres` role manually using `createuser -P -d postgres`. The command appears in the digital ocean guide link but the fact that the role is named `postgres` and the password is `password` is specific to the application.

From https://github.com/rubyforgood/human-essentials/blob/main/config/database.yml

Author: @heyapricot do you think the removal of the `password` requirement locally for the DB will avoid this issue? If so, I'am not sure what the upside is for having a password locally.

Reviewer 3: I'll do a clean setup with a modified version of config/database.yml (with the password removed) and see how that goes

###### Ubuntu

> TODO

Reviewer 2: Don't we have ubuntu instructions here: https://github.com/rubyforgood/human-essentials/blob/main/installation.md

Or are they out of date?

Author: @seanmarcia I'am not sure. We do have a Ubuntu user that I can double check with.

Reviewer 5: Hey, I'm the Ubuntu user ;)

The setup is basically working and up-to-date, just got a few remarks:

- I did not install the packages mentioned in the beginning, as I already had them installed.
- Running sudo npm -g install yarn did not work ("command not found"), just npm -g install yarn did the trick.
- I skipped over the rvm part, as I am using rbenv.  I'd probably put a message there for people to double-check, whether they already have another version manager installed, as for example rbenv and rvm _will_ clash with one-another.
- The part where you explain how to set up the environment variables wasn't really clear to me. I just added them to my root .bashrc in the end and set it as source inside the project folder, which worked just fine.
- Running rails db:setup db:migrate db:test:prepare db:seed did cause an issue for the first time around during DB seeding. Something about "address1" not being a valid Active Model attribute or something. Had to restart my machine due to something unrelated, though, and the next time around the command ran just fine and I am no longer able to reproduce the problem now.

Project seems to be running just fine on my Ubuntu machine now and the setup went pretty quick all in all.

Author: Thanks for the feedback @florianfelsing! I had a follow up question for:
> The part where you explain how to set up the environment variables wasn't really clear to me. I just added them to my root .bashrc in the end and set it as source inside the project folder, which worked just fine.

The goal was to allow the `bin/setup` command to setup your .env file which would be used. Was that unclear and how could we be clearer about that?

Reviewer 5: > Thanks for the feedback @florianfelsing! I had a follow up question for:
> 
> > The part where you explain how to set up the environment variables wasn't really clear to me. I just added them to my root .bashrc in the end and set it as source inside the project folder, which worked just fine.
> 
> The goal was to allow the `bin/setup` command to setup your .env file which would be used. Was that unclear and how could we be clearer about that?

I overlooked that part of the process since it wasn't mentioned in the Ubuntu-specific installation guide. Only realized too late that it was to be found in the README.

Also as for what @divagueame wrote, this is exactly the error that I encountered once and that disappeared once I had to reboot my system.

###### Windows

> TODO

Reviewer 2: Maybe writes something like:
Do you develop on windows? We'd love to hear (and for you to submit a PR explaining) how you do it.

Author: Love it! I'll include it here.

> TODO

Author: ```suggestion
Do you develop on windows? We'd love to hear (and for you to submit a PR explaining) how you do it. 🙏🏻
```

After completing this step, you should have both ruby, node, and postgres installed. Now you can proceed to the next step which is to setup the application.

### Run Setup
Run the following command should run all the neccessary steps to get up and running.
```

> bin/setup

Reviewer 3: Following the guide there is no mention of installing yarn. When I ran the `bin/setup`  script I got an error related to db/seed.

After I installed yarn (Classic yarn because I couldn't get yarn working with corepack for v12) the bin/setup script ran without trouble

Author: Nice catch! Would you mind adding a step to install yarn prior to this step?

Side note - Argh! I can't wait to remove node dependency!

Reviewer 3: Sure I'll do it! (I won't be able to add it before friday though)

Reviewer 3: Solved by https://github.com/rubyforgood/human-essentials/pull/2970/commits/81eddeb85b3afd6857b63a3085500b5809d5c945

```

### Run the application
Run the following command to start the server locally.
```
bin/start

> ```

Reviewer 3: Got the following errors after running bin/start
![error](https://user-images.githubusercontent.com/14355495/173702064-13ec52de-3c91-4e30-859e-5c14b09abbac.png)
<img width="1792" alt="Screen Shot 2022-06-14 at 18 48 08" src="https://user-images.githubusercontent.com/14355495/173702074-65c7d226-890b-4810-a4d3-7abc6436a252.png">

Reviewer 3: This solved it https://github.com/rails/webpacker/issues/2825#issuecomment-764888322

Author: Do you think adding that `rails webpacker:compile` step in the `bin/setup` would work?

Reviewer 3: Hopefully that will work. I'll give it a shot and let you know how it went
